mlptrain/training/selection.py

```diff
@@ -187,3 +197,101 @@ def n_backtrack(self) -> int:
     def _n_training_envs(self) -> int:
         """Number of training environments available"""
         return len(self._k_vec)
+
+def outlier_identifier (configuration: 'mlptrain.Configuration',
+                        configurations:'mlptrain.ConfigurationSet',
+                        dim_reduction: bool = False,
+                        distance_metric: str = 'euclidean',
+                        n_neighbors: int = 15) -> int:
+    """
+    This function aimed to identify whether a new data (configuation)
+    is the outlier for the existed data (configurations) by Local Outlier 
+    Factor (LOF). For more details about LOF method, please see the lit. 
+    Breunig, M. M., Kriegel, H.-P., Ng, R. T. & Sander, J. LOF: Identifying 
+    density-based local outliers. SIGMOD Rec. 29, 93–104 (2000).
+
+    -----------------------------------------------------------------------
+    Arguments:
+    
+    dim_reduction: if Ture, dimensional reduction (PCA applied here) will
+                   be performed before LOF calculation.
+    distance_metric: distance metric used in LOF,
+                     which could be one of 'euclidean', 
+                     'cosine' and 'manhattan’.
+    n_neighbors: how many neighbors should be consider when compute LOF.
+
+    -----------------------------------------------------------------------
+    Returns:
+    
+    -1 for anomalies/outliers and +1 for inliers.
+    """
+
+    m1 = soap_matrix(configurations)
+    m1 /= np.linalg.norm(m1, axis=1).reshape(len(configurations), 1)
+
+    v1 = soap_matrix(configuration)
+    v1 /= np.linalg.norm(v1, axis=1).reshape(1, -1)
+
+    if dim_reduction:
+        pca = PCA(n_components=3)
+        m1 = pca.fit_transform(m1)
+        v1 = pca.transform(v1)
+
+    clf = LocalOutlierFactor(n_neighbors=n_neighbors, metric=distance_metric, novelty=True, contamination=0.2)
+    'contamination: define the porpotional of outliner in the data, the higher, the less abnormal'
+                 
+    clf.fit(m1)
+
+    new = clf.predict(v1)
+
+    return new
+
+class AtomicEnvDistance(SelectionMethod):
+    def __init__(self,
+                 pca: bool = False,
+                 distance_metric: str = "euclidean",
+                 n_neighbors: int = 15):
+        """
+        Selection criteria based on access whether the configuration is 
+        outlier by outlier_identifier function
+        -----------------------------------------------------------------------
+        Arguments:
+            pca: whether to do dimenstional reduction by PCA. 
+                 Because the selected distance_metric may potentially suffer from 
+                 the curse of dimensionality, the dimensionality reduction step 
+                 (using PCA) could be applied before calculating the LOF. 
+                 This would ensure good performance in high-dimensional data space.
+            the other arguments, please see details in outlier_identifier function
```

Comment:
```suggestion
        Selection criteria based on analysis whether the configuration is 
        outlier by outlier_identifier function
        -----------------------------------------------------------------------
        Arguments:
            pca: whether to do dimensionality reduction by PCA. 
                 As the selected distance_metric may potentially suffer from 
                 the curse of dimensionality, the dimensionality reduction step 
                 (using PCA) could be applied before calculating the LOF. 
                 This would ensure good performance in high-dimensional data space.
            For the other arguments, please see details in the outlier_identifier function
```